src/unix/linux/disk.rs

```diff
@@ -83,39 +83,55 @@ impl DiskInner {
         self.is_read_only
     }
 
-    pub(crate) fn refresh(&mut self) -> bool {
-        self.efficient_refresh(&disk_stats())
+    pub(crate) fn refresh_specifics(&mut self, refresh_kind: DiskRefreshKind) -> bool {
+        self.efficient_refresh(refresh_kind, &disk_stats(&refresh_kind))
     }
 
-    fn efficient_refresh(&mut self, procfs_disk_stats: &HashMap<String, DiskStat>) -> bool {
-        let Some((read_bytes, written_bytes)) =
-            procfs_disk_stats.get(&self.actual_device_name).map(|stat| {
-                (
-                    stat.sectors_read * SECTOR_SIZE,
-                    stat.sectors_written * SECTOR_SIZE,
-                )
-            })
-        else {
-            sysinfo_debug!("Failed to update disk i/o stats");
-            return false;
-        };
+    fn efficient_refresh(
+        &mut self,
+        refresh_kind: DiskRefreshKind,
+        procfs_disk_stats: &HashMap<String, DiskStat>,
+    ) -> bool {
+        if refresh_kind.kind() && self.type_ == DiskKind::Unknown(-1) {
+            self.type_ = find_type_for_device_name(&self.device_name);
+        }
 
-        self.old_read_bytes = self.read_bytes;
-        self.old_written_bytes = self.written_bytes;
-        self.read_bytes = read_bytes;
-        self.written_bytes = written_bytes;
-
-        unsafe {
-            let mut stat: statvfs = mem::zeroed();
-            let mount_point_cpath = to_cpath(&self.mount_point);
-            if retry_eintr!(statvfs(mount_point_cpath.as_ptr() as *const _, &mut stat)) == 0 {
-                let tmp = cast!(stat.f_bsize).saturating_mul(cast!(stat.f_bavail));
-                self.available_space = cast!(tmp);
-                true
+        if refresh_kind.io_usage() {
+            if self.actual_device_name.is_none() {
+                self.actual_device_name = Some(get_actual_device_name(&self.device_name));
+            }
+            let (read_bytes, written_bytes) = if let Some((read_bytes, written_bytes)) =
+                procfs_disk_stats
+                    .get(self.actual_device_name.as_ref().unwrap())
+                    .map(|stat| {
+                        (
+                            stat.sectors_read * SECTOR_SIZE,
+                            stat.sectors_written * SECTOR_SIZE,
+                        )
+                    }) {
+                (read_bytes, written_bytes)
             } else {
-                false
+                sysinfo_debug!("Failed to update disk i/o stats");
+                (0, 0)
+            };
+
+            self.old_read_bytes = self.read_bytes;
```

Comment:
Should only be set if `statvfs` succeeded.

Reply:
Done